### ISSUE TITLE
Cannot delete clients

### ISSUE THREAD

**Issue by HumanNr4584093104** [te-online/timemanager#53]
Installed the android app, added two clients that never showed up in the nc app. Tried inputting both several times, nothing.

Log on to Nextcloud, see clients:

```
My project
Housekeeping
Housekeeping
My project
My project
My project
```

Tried to delete a client, confirm deletion, page reloads - Nothing has been deleted. If I rename a project it's duplicated

```
My project
Housekeeping
Housekeeping
Housekeeping (Renamed)
My project
My project
My project
My project (Renamed)
```

Reinstalled the android app and try to sync, doesn't find any clients.

Add a client manually on the web-interface. App doesn't detect it.

**Comment by te-online** [comment]
Thanks for reporting! 🙌  

On first glance, this sounds like some issue with your setup. I can't reproduce what you describe on my personal Nextcloud instance, for example. However, I haven't been using the Android app for a while, so can't say anything about that.

Any chance you can provide some more details about your environment? 


<details>
<summary>Those details would be helpful 👇 </summary>

- Nextcloud version
- PHP version
- Server OS
- MySQL/MariaDb version
- Webserver (Apache/Nginx)
- Android version
- Android app version

</details>

I'd also encourage you to find out if you can enable some sort of debug mode in the Android app to have some insight into what is happening on that side of things: [Gitlab repository](https://gitlab.com/joleaf/nc-timemanager-mobile-app)

**Comment by kkatalinux** [comment]
Hello! I have the same issue like HumanNr4584093104. 
I added two clients in the Android app, they showed up in the NC app several times (one client 28 times, the other one 53 times)s, and now it is impossible to delete or edit them. I tried to delete the folders under the corresponding NC folder structure on the server after removing the app, then reinstalled it, I did this several times but these clients seem to survive everything :) Impossible to delete or edit them. 
Nextcloud 22.2.0

In the Android app everything seems to be okay, I can add and remove clients and projects and tasks, they show up in the NC app. 

Maybe there is a secret folder where these data are stored and did not get deleted? Can you give me a hint where I should look to find them?
Many thanks!
![timemanager_clients](https://user-images.githubusercontent.com/55601005/139670305-eb7a05da-b680-42af-8f64-a0bb31accd7f.png)

**Comment by HumanNr4584093104** [comment]
@kkatalinux I would think the listings are in MySQL/MariaDB tables that are separate from the folders. My random guess is that there is an issue with the way the app adds/remove clients and the long term fix is with the phone app developer. But removing invalid entries in the Nextcloud app is something I hope we will be able to do through the web interface.

I am currently working on figuring out how to provide all the requested details to the developer (see their original reply), I suggest you do the same and we'll give the dev the best chance of finding the issue! 👍

**Comment by te-online** [comment]
I have a feeling this might be related to fresh setups and a recent bug I fixed in the storage service. Just a theory for now, but I will make sure to look into this during the weekend. That being said, it'd be a great help if both of you could inspect the Nextcloud log (Settings -> Log, in your Nextcloud browser UI) for any errors potentially related to this.

**Comment by HumanNr4584093104** [comment]
Yes, I get errors. I just cut and pasted from the weblog, do you need the content of the raw log?

My server deets: Debian NextcloudPi on ARM64 Pi4
Operating system: Linux 5.10.43
Webserver: Apache (fpm-fcgi)
Database: mysql 10.3.31
PHP version: 7.3.31-1~deb10u1
Nextcloud version: 21.0.4 - 21.0.4.1

Phone: Android11 / LineageOS 18.1
App: Latest app from Play Store


```
Error	PHP	Error: Undefined property: OCA\TimeManager\Db\Client::$hours at /var/www/nextcloud/apps/timemanager/templates/projects.php#92

    /var/www/nextcloud/apps/timemanager/templates/projects.php - line 92:

    OC\Log\ErrorHandler::onError(8, "Undefined p ... s", "/var/www/ne ... p", 92, { 0: "And 10 ... }})

    /var/www/nextcloud/lib/private/Template/Base.php - line 180:

    include("/var/www/ne ... p")

    /var/www/nextcloud/lib/private/Template/Base.php - line 150:

    OC\Template\Base->load("/var/www/ne ... p", { 0: "And 2 ... "})

    /var/www/nextcloud/lib/private/legacy/OC_Template.php - line 180:

    OC\Template\Base->fetchPage({ 0: "And 2 ... "})

    /var/www/nextcloud/lib/public/AppFramework/Http/TemplateResponse.php - line 210:

    OC_Template->fetchPage({ 0: "And 2 ... "})

    /var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php - line 172:

    OCP\AppFramework\Http\TemplateResponse->render()

    /var/www/nextcloud/lib/private/AppFramework/App.php - line 157:

    OC\AppFramework\Http\Dispatcher->dispatch(OCA\TimeMana ... {}, "projects")

    /var/www/nextcloud/lib/private/Route/Router.php - line 302:

    OC\AppFramework\App::main("OCA\\Timema ... r", "projects", OC\AppFramew ... {}, { _route: "t ... "})

    /var/www/nextcloud/lib/base.php - line 993:

    OC\Route\Router->match("/apps/timemanager/projects")

    /var/www/nextcloud/index.php - line 37:

    OC::handleRequest()
```

**Comment by te-online** [comment]
Hey @HumanNr4584093104 – the error you're seeing in your log seems to be a template error. Maybe a corrupted project in your database? It seems unlikely this is related to the other issues / the sync issue. I've tried to address the sync issue on `master` by assuming that the Android app doesn't handle sync conflicts. 

Do you have any possibility to try current `master` on a testsystem (please don't test with production data). 

I'm working on writing some unit tests, so in the future I can be sure not to break sync functionality with any changes or fixes.

**Comment by te-online** [comment]
@kkatalinux What you describe sounds very much like the Android app doesn't do conflict handling and re-creates objects every time it syncs. Does this happen when you make changes in the Android app that are only being synced after you've made changes in the Nextcloud app? I don't have an Android phone to test right now, so I'll have to get an emulator with Playstore up & running to re-create the issue. This will take a little more time than I thought 🙁

**Comment by te-online** [comment]
_Update:_ I spent 1 hour fiddling with the Android emulator to get it to sign me into my local dev nextcloud, but that won't work 😢 So, I ended up setting up the app with a Nextcloud instance on the internet, running latest TimeManager version 0.2.5. And here I can reproduce the issue. Objects are re-created over and over. The reason why you can't seem to edit those items is that they all share the same `uuid` 🤯 (that should have been a unique key?!). 

@joleaf can you maybe help us out here 🙏 Could you test current `master` with your app and see if this fixes the duplicate objects after sync issue? Or can you say why this is happening? I can't see any recent code changes in the Android app, so I'm assuming it's something I did that broke syncing?

**Comment by joleaf** [comment]
oh s**** 🗡️
At the moment I have deadlines at work so I can not spend much time on this.
But I think it has to do something with the UUID.. If the android app creates a new entry, it already sets the uuid of this object. So maybe this is something the TM handles now different with the API endpoint?

I can publish a new android version with some flutter updates, but I don't think, that this fixes the issue..

**Comment by te-online** [comment]
Hey everyone, I just released a new version 🎉  

However, my testing hasn't really been conclusive regarding the Android app 😩 

I was able to sync successfully with the updated backend, using my emulator version of the app.

For what I assume was or is the issue, read my [comments in the changelog here](https://github.com/te-online/timemanager/tree/v0.2.6#026-beta-13-november-2021).

_One thing_ I'm pretty certain about is that 0.2.6 will fix the symptoms described in this issue. Your objects will be de-duplicated by the database migration included in this release and you will be able to delete and edit each object individually again. There's also a unique-index on the columns now, so every insertion of duplicate objects will hard-fail in the future and prevent this issue.
